### DISCUSSION TITLE
[보안] Access 토큰과 Refresh Token 관리에 대한 제안사항 공유 드립니다.

### DISCUSSION THREAD

**Discussion by Starlight258** [Attica-org/athens-backend#52] [General]
#51 
안녕하세요 팀원 여러분 🙋‍♀️
현재 프로젝트에 `액세스` 토큰과 `리프레시` 토큰을 모두 쿠키에 저장하고 있는데, 제안 사항이 있어 글을 작성하게 되었습니다.

1. 일반적으로 액세스 토큰을 메모리에 저장하는 것을 권장한다고 합니다. 저희 프로젝트에서 **둘 다 쿠키로 저장하기로 결정한 이유**가 있을까요?
혹시 고려한 사항이 있다면 공유 부탁드려요. 

#### access token을 메모리에 저장할 때의 장점
- access token은 짧은 유효기간을 가지므로 브라우저 탭을 전환하거나 페이지를 새로고침하면 사라진다. 이럴 경우 새로 access token을 발급받아야 하는데, 메모리에 저장하면 필요할 때마다 refresh token을 이용해 새 access token을 쉽게 발급받을 수 있다.
- 메모리 저장 시 access token 크기 제한이나 Authorization 헤더 사용 등의 제약이 없다.

자세한 내용은 아래 글에서 보실 수 있습니다.
[JWT Tokens Securely in the Front-End](https://codeburst.io/localstorage-vs-cookies-all-you-need-to-know-about-storing-jwt-tokens-securely-in-the-front-end-70dc0a9b3ad3)

[보안을 위한 JWT 진화](https://www.youtube.com/watch?v=Q4c0n83S1xs&list=PLJkjrxxiBSFATow4HY2qr5wLvXM6Rg-BM&index=2)
<br/>

2. 현재 액세스 토큰과 리프레시 토큰을 쿠키에 저장하고 있어 `CSRF` 공격에 취약할 수 있습니다. `mvp2`에서는 `Anti-CSRF` 토큰을 포함하는 식으로 적절한 보안 조치를 취하는 것이 어떨까요? 

관련해서 액세스 토큰과 리프레시 토큰을 함께 저장할 경우 둘 다 탈취되어 위험이 증가한다는 글도 보았는데 링크 첨부하겠습니다.
[Is storing both Access and Refresh tokens in httpOnly cookie kinda pointless](https://www.reddit.com/r/reactjs/comments/z7lmo0/is_storing_both_access_and_refresh_tokens_in/?onetap_auto=true&one_tap=true)

3. 모든 요청에 액세스 토큰과 리프레시 토큰이 쿠키로 함께 전송되고 있는데, 리프레시 토큰은 재발급 요청 시에만 전송하고, 다른 요청에는 액세스 토큰만 포함시키는 것이 좋을 것 같아요. **프론트** 구현시 `SameSite` 속성을 `strict`로 설정하여 방지하는 것을 제안드립니다.

4. 액세스 토큰과 리프레시 `토큰의 유효기간`을 짧게 설정하는 것이 좋을 것 같아요. 예를 들어 액세스 토큰은 30초~15분, 리프레시 토큰은 하루 정도로 설정하는 것을 제안드립니다.

@H-sooyeon @big-cir @yybmion

**Comment by big-cir** [comment]
로그인 시 access, refresh를 발급해주고 access를 재발급 할때만 다시 쿠키로 refresh를 전송하기로 저번에 얘기하지 않았었나요???

**Reply by Starlight258** [reply]
@big-cir 네 현재 쿠키가 /로 저장되어있어서 모든 요청에 대해 전달되도록 구현되어있어 말씀드렸습니다.
제안 드린 사항은 access 토큰은 보통 응답 본문이나 localStorage에 저장되는데 저희 프로젝트에는 쿠키에 저장되어 있어 디스커션 열었어요.

**Reply by big-cir** [reply]
잘 이해를 못했는데 조금 풀어서 설명 해주실 수 있나요?

**Reply by Starlight258** [reply]
PR 보시면 쿠키 생성할때 setPath가 전체 경로로 되어있어요.
어떤 부분이 이해가 안가시는지 이야기해주시면 아는 부분까지 설명해드릴게요. 🙌

**Reply by big-cir** [reply]
말씀해주신 부분 이해했습니다! 감사합니다 :)

**Comment by H-sooyeon** [comment]
보고 있는데, 
일단 3번은 토큰을 서버에서 쿠키에 담아주실 때 SameSite 설정 하시면 됩니다.
쿠키에 담아주실 땐 프론트에서 자바스크립트 실행이 불가능하기 때문에 쿠키에 접근할 수 없습니다.
-> 쿠키 값 뽑기, 쿠키 설정 변경 등 모든 자바스크립트 실행이 불가능합니다.

**Reply by Starlight258** [reply]
그렇군요! 쿠키를 서버에서 생성하므로 Samesite 설정까지 서버에서 해주어야하네요  !

**Comment by H-sooyeon** [comment]
> **access token을 메모리에 저장할 때의 장점**
> access token은 짧은 유효기간을 가지므로 브라우저 탭을 전환하거나 페이지를 새로고침하면 사라진다. 이럴 경우 새로 access token을 발급받아야 하는데, 메모리에 저장하면 필요할 때마다 refresh token을 이용해 새 access token을 쉽게 발급받을 수 있다.
> 메모리 저장 시 access token 크기 제한이나 Authorization 헤더 사용 등의 제약이 없다.

쿠키보다 localStorage가 메모리 크기가 더 크긴 합니다.
위의 장점은 쿠키에도 적용되기 때문에 로컬스토리지와 쿠키의 차이라고 하기엔 어려울 것 같아요!
로컬스토리지에 영구적으로 access token을 저장해놓을게 아니고 유효기간 동안만 저장되도록 한다면 쿠키와 동일하다는 의미입니다.

+새로고침 or 탭 전환 시 사라지는건 유효기간이 끝났을 때의 이야기입니다.
유효기간이 끝나지 않았다면 데이터가 그대로 저장되어 있습니다.

**Reply by Starlight258** [reply]
그렇군요! localStorage에 저장한다고 딱히 장점이 있는것은 아니네요 ! 😃

access 토큰을 refresh 토큰과 함께 저장하는 것이 위험해서 따로 저장한다는 의견에 대해서는 어떻게 생각하시나요?
[Is storing both Access and Refresh tokens in httpOnly cookie kinda pointless](https://www.reddit.com/r/reactjs/comments/z7lmo0/is_storing_both_access_and_refresh_tokens_in/?onetap_auto=true&one_tap=true)

관련 블로그나 영상에서도 access token은 메모리에 저장하는 것을 추천하는데,, 다들 둘 다 쿠키로 저장하지 않는 이유가 있는것인지 궁금하네요 ! 분명 이유가 있을텐데요..
[JWT Tokens Securely in the Front-End](https://codeburst.io/localstorage-vs-cookies-all-you-need-to-know-about-storing-jwt-tokens-securely-in-the-front-end-70dc0a9b3ad3)
[보안을 위한 JWT 진화](https://www.youtube.com/watch?v=Q4c0n83S1xs&list=PLJkjrxxiBSFATow4HY2qr5wLvXM6Rg-BM&index=2)

**Reply by H-sooyeon** [reply]
방법은 다양합니다.

아래 링크 보시면 될 것 같네요.
https://www.rfc-editor.org/rfc/rfc6819#section-4.1.3

위 링크는 올려주셨던 링크 중 하나의 답글에서 볼 수 있습니다.

> 관련해서 액세스 토큰과 리프레시 토큰을 함께 저장할 경우 둘 다 탈취되어 위험이 증가한다는 글도 보았는데 링크 첨부하겠습니다.
> [Is storing both Access and Refresh tokens in httpOnly cookie kinda pointless](https://www.reddit.com/r/reactjs/comments/z7lmo0/is_storing_both_access_and_refresh_tokens_in/?onetap_auto=true&one_tap=true)

**Reply by Starlight258** [reply]
> 방법은 다양합니다.
> 
> 아래 링크 보시면 될 것 같네요. https://www.rfc-editor.org/rfc/rfc6819#section-4.1.3

여기 글에는 액세스 토큰을 일시적으로 메모리에 저장하던지, 리프레시 토큰을 액세스 토큰과 동일한 보호 수준으로(토큰 로테이션, 발급 제한등) 보호하라고 되어있어어서 오히려 메모리에 액세스 토큰을 저장하는 근거가 될 것 같아요.

**Reply by H-sooyeon** [reply]
아 죄송합니다. 제가 반대로 해석했나 보네요.

refresh token은 유효기간이 access보단 길기 때문에 보호를 위해 쿠키(http Only, secure 등 설정)에 저장하는 방식이 맞는 것 같고,
쿠키 탈취로 access + refresh 토큰 모두 접근할 수 있게 되는 문제 해결을 위해선
access token을 쿠키가 아닌 방식으로 받되 전역 상태로 프론트에서 보관하면서 api로 전달드리는 방법이 나을 것 같다고 저는 생각합니다.

메모리(상태로 저장)가 아닌 로컬/쿠키 둘 중 하나로 토큰을 보관한다고 하면 로컬 스토리지는 쿠키보다 보안에 취약하기 때문에 부정적인 의견입니다. (httpOnly 등의 설정이 불가능합니다.)
상태로 저장하게 되면 확실히 처음 올려주신 글처럼 새로고침 시 데이터가 모두 날아가기 때문에 서버로 access token을 요청하는 트래픽이 기존보다 늘어나긴 합니다.

**Reply by Starlight258** [reply]
넵 저도 보내주신 [RFC 6819-OAuth 2.0 보안 고려사항](https://www.rfc-editor.org/rfc/rfc6819#section-4.1.3)에 따라서 클라이언트에서 액세스 토큰을 메모리에 저장하고 액세스 토큰은 헤더나 쿠키가 아닌 response body로 전달하는 방법이 좋을 것 같아요. 
OAuth도 액세스 토큰과 리프레시 토큰을 사용하니 충분히 가이드라인이 된다고 생각합니다.

**Comment by big-cir** [comment]
> 액세스 토큰과 리프레시 토큰의 유효기간을 짧게 설정하는 것이 좋을 것 같아요. 예를 들어 액세스 토큰은 30초~15분, 리프레시 토큰은 하루 정도로 설정하는 것을 제안드립니다.

만료시간을 이렇게 짧게 가져가는 이유가 있나요..?

**Reply by Starlight258** [reply]
토큰 탈취 위험때문이에요. 제가 본 예시에서는 30초~15분인데 너무 짧다면 30분 정도도 괜찮아보여요.
reddit에서는 3분을 넘지 말라는 이야기도 있네요.
[Is storing both Access and Refresh tokens in httpOnly cookie kinda pointless](https://www.reddit.com/r/reactjs/comments/z7lmo0/is_storing_both_access_and_refresh_tokens_in/?onetap_auto=true&one_tap=true)

**Reply by big-cir** [reply]
accessToken의 유효기간 설정을 판단 기준이 없어서 어렵네요.. ㅎㅎ
하지만 단순히 저 글만 보고 결정 하기에는 근거가 조금 부족하다는 생각이 들긴합니다.!
어떤 주기가 서비스에 적합한지 판단하기 힘들면 일반적으로 어느 정도 주기를 가지는지 찾아봐야겠습니다.

**Comment by H-sooyeon** [comment]
토이 프로젝트를 진행할 땐 펀의성을 위해 로컬스토리지에 저장하는 경우가 많습니다.
하지만 회의 때 공유했던 XSS 공격 등의 문제로 쿠키로 httpOnly 설정으로 사용하는 경우가 많은 것 같아요.
지금 커리어리의 access token 보관 방법에 대한 질문 글을 보고 있는데 지금까지 본 답변은 다 쿠키에 httpOnly 설정을 두고 사용하라고 합니다.

일단 계속 봐보고 다른 답변이 있으면 공유하겠습니다.

**Reply by Starlight258** [reply]
커리어리 글 첨부해주시면 저도 읽어볼게요!

**Reply by H-sooyeon** [reply]
커리어리 현직자 Q&A에서 access token 검색하시면 보실 수 있어요!
그 중 하나 올려놓겠습니당

[커리어리 access token 보관 질문](https://careerly.co.kr/qnas/5944)

**Reply by Starlight258** [reply]
감사합니다! 글 보고 있는데, nextJs에서는 쿠키와 달리 헤더에서 토큰을 추출하려면 추가 작업이 필요하나요?

**Reply by H-sooyeon** [reply]
특별한 작업이 필요하진 않습니다.

응답 메시지로 토큰을 받게될 시 스토리지에 토큰 보관 과정은 아래와 같습니다.
1. 응답 헤더 or 바디로 토큰을 받음
2. 응답 메시지에서 토큰 추출
3. 스토리지에 토큰 저장

스토리지에서 서버로 전달 시
1. 스토리지에서 토큰 추출
2. 요청 메시지에 추출한 토큰 추가
3. 서버로 전달

대략 간단하게 위와 같은 방식으로 이루어집니다.

**Reply by H-sooyeon** [reply]
+쿠키는 위의 과정 필요 없이 알아서 메시지에 담겨 서버로 전달됩니다.

**Reply by Starlight258** [reply]
아~ 따로 추가 작업은 필요 없는 거군요

**Comment by yybmion** [comment]
쿠키로 계속 저장한다면 Oauth2에서 제한사항이 걸리나요?

계속 찾아보아도 access 토큰을 보통은 localStorage에 저장하지만 쿠키에 저장하지 않을 이유도 없다고 생각하는게 두 부분다 장점과 단점이 있어 뭐가 더 낫다고 생각이 들지가 않네요.

하지만 한편으로는 사람이 보통 많이 사용하는(Access localstorage 저장)이유가 있다고는 생각합니다...
